File: cuts/priority_queue.cpp
// Priority queue implementation in C++ using heapq
// Avaliable methods are: 
// void Enqueue(string obj, int priority)
// std::string Dequeue()
// The bigger the priority number, the higher the priority
// On the same priority will be used FIFO


class PriorityQueue {
private:
    std::priority_queue<std::tuple<int, int, std::string>> queue;
    int counter;
public:
    PriorityQueue() : counter(0) {}

    void Enqueue(std::string obj, int priority) {
        queue.push(std::make_tuple(priority, counter++, obj));
    }

    std::string Dequeue() {
        if (queue.empty()) {
            throw std::runtime_error("Dequeue from an empty priority queue");
        }
        std::string result = std::get<2>(queue.top());
        queue.pop();
        return result;
    }
};


class PriorityQueue2 {
private:
    std::vector<std::tuple<int, int, std::string>> queue;
    int counter;

    int FindHighestPriorityIndex() {
        if (queue.empty()) {
            throw std::runtime_error("Dequeue from an empty priority queue");
        }

        int highestPriorityIndex = 0;
        for (int i = 1; i < queue.size(); ++i) {
            if (std::get<0>(queue[i]) > std::get<0>(queue[highestPriorityIndex]) || 
                (std::get<0>(queue[i]) == std::get<0>(queue[highestPriorityIndex]) &&
                std::get<1>(queue[i]) < std::get<1>(queue[highestPriorityIndex]))) {
                highestPriorityIndex = i;
            }
        }

        return highestPriorityIndex;
    }
public:
    PriorityQueue2() : counter(0) {}

    void Enqueue(std::string obj, int priority) {
        queue.push_back(std::make_tuple(priority, counter++, obj));
    }

    std::string Dequeue() {
        int highestPriorityIndex = FindHighestPriorityIndex();
        std::string obj = std::get<2>(queue[highestPriorityIndex]);
        queue.erase(queue.begin() + highestPriorityIndex);
        return obj;
    }
}

File: cuts/priority_queue.py
# Priority queue implementation in Python using heapq
# Avaliable methods are: 
# def Enqueue(obj: Object, priority: int) -> None
# def Dequeue() -> Object
# The bigger the priority number, the higher the priority
# On the same priority will be used FIFO

import heapq
from typing import Any

class PriorityQueue:
    def __init__(self):
        self.queue = []
        self.counter = 0
    
    def Enqueue(self, obj: Any, priority: int) -> None:
        heapq.heappush(self.queue, (-priority, self.counter, obj))
        self.counter += 1

    def Dequeue(self) -> Any:
        if len(self.queue) == 0:
            return None
        return heapq.heappop(self.queue)[2]
    

class PriorityQueue2:
    def __init__(self):
        self.queue = []
        self.counter = 0

    def __findHighestPriority(self):
        highest = 0
        for i in range(1, len(self.queue)):
            if self.queue[i][0] > self.queue[highest][0] or \
                (self.queue[i][0] == self.queue[highest][0] and \
                    self.queue[i][1] < self.queue[highest][1]):
                highest = i
        return highest

    def Enqueue(self, obj: Any, priority: int) -> None:
        self.queue.append((priority, self.counter, obj))
        self.counter += 1

    def Dequeue(self) -> Any:
        if len(self.queue) == 0:
            return None

        highest = self.__findHighestPriority()
        _, _, obj = self.queue.pop(highest)
        return obj
